File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.java
class Solution {
    
    class Union {
        Map<Integer, Integer> parent;
        Map<Integer, Integer> rank;
        int maxSize;

        public Union() {
            parent = new HashMap<>();
            rank = new HashMap<>();
            maxSize = 1;
        }

        private int find(int num) {
            if (parent.get(num) != num) {
                parent.put(num, find(parent.get(num)));
            }
            return parent.get(num); 
        }

        public void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);

            if (rootX != rootY) {
                int rankX = rank.get(rootX);
                int rankY = rank.get(rootY);

                if (rankX > rankY) {
                    parent.put(rootY, rootX);
                    rank.put(rootX, rankX + rankY);
                    maxSize = Math.max(maxSize, rank.get(rootX));
                } else {
                    parent.put(rootX, rootY);
                    rank.put(rootY, rankX + rankY);
                    maxSize = Math.max(maxSize, rank.get(rootY));
                }
            }
        }

        public void add(int x) {
            parent.put(x, x);
            rank.put(x, 1);
        }

        public boolean contains(int x) {
            return parent.containsKey(x);
        }
        

        public int getMax() {
            return maxSize;
        }
    }


    public int longestConsecutive(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        Union u = new Union();
        for (int num : nums) {
            if (!u.contains(num)) {
                u.add(num);
                
            } 
        }

        for (int num : nums) {
            if (u.contains(num + 1)) {
                u.union(num, num + 1);
            }
        }

        return u.getMax();
    }
}

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
# class UnionFind:
#     def __init__(self):
#         self.parents = {}
#         self.counts = {}
#         self.max = 1

#     def find(self, u):
#         if not u in self.parents:
#             self.parents[u] = u
#             self.counts[u] = 1
#             return u
        
#         if self.parents[u] != u:
#             self.parents[u] = self.find(self.parents[u])

#         return self.parents[u]

#     def join(self, u, v):
#         rootU = self.find(u)
#         rootV = self.find(v)

#         if rootU != rootV:
#             if self.counts[rootU] > self.counts[rootV]:  # Merge smaller into larger
#                 rootU, rootV = rootV, rootU

#             self.parents[rootU] = rootV
#             self.counts[rootV] += self.counts[rootU]
#             self.max = max(self.max, self.counts[rootV])
            


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # if len(nums) == 0:
        #     return 0

        # numSet = set(nums)

        # union = UnionFind()

        # for num in numSet:
        #     if num - 1 in numSet:
                
        #         union.join(num, num - 1)
            

        # return union.max

        numSet = set(nums)
        maxLen = 0

        for num in numSet:
            if not num - 1 in numSet:
                currNum = num
                currLen = 1

                while currNum + 1 in numSet:
                    currNum += 1
                    currLen += 1

                maxLen = max(currLen, maxLen)

        return maxLen
